tests/functional_test_base.go

```diff
@@ -169,10 +168,6 @@ func (s *FunctionalTestBase) setupSuite(defaultClusterConfigFile string, options
 		s.archivalNamespace = s.randomizeStr("functional-archival-enabled-namespace")
 		s.Require().NoError(s.registerArchivalNamespace(s.archivalNamespace))
 	}
-
-	// For tests using SQL visibility, we need to wait for search attributes to be available as part of the ns config
-	// TODO: remove after https://github.com/temporalio/temporal/issues/4017 is resolved
-	time.Sleep(2 * NamespaceCacheRefreshInterval)
```

Comment:
Maybe this should be replaced with an eventually condition that checks whether the expected data has been propagated? Just want to be very sure this change won't cause more tests to flake.

Reply:
Looks like I also needed to add this DC override:

```
		// Better to read through in tests than add artificial sleeps (which is what we previously had).
		dynamicconfig.ForceSearchAttributesCacheRefreshOnRead.Key(): true,
```

Reply:
Yeah, just wanted to double check because I've tried to remove these and similar statements in the past and it caused seemingly unrelated flakes.